### PR TITLE
feat: add builtin uninstaller as alternative to calling pip - variant without rollback

### PR DESCRIPTION
**Pull request by radoering**
# Pull Request Check List

Closes: #10904

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

Introduce a builtin uninstaller (adapted from pip's code, dropping legacy handling and the option to rollback) so that we do not have to run a pip subprocess for each uninstall. This makes uninstalls (and updates) faster. The new builtin uninstaller is not used by default. You have to set `installer.builtin-uninstall` to use it. (We can switch it later and remove the setting even later, just as we did with `installer.modern-installation`.)

Alternative to #10904 without the option to rollback uninstalls as proposed in https://github.com/python-poetry/poetry/pull/10904#discussion_r3326030326.

I started from #10904 removing the rollback mechansim, which resulted in https://github.com/radoering/poetry/commit/0d1d42e3e60b5d8a4bfb5cabc2585750176398a2. (This is another alternative on another branch.) This still keeps the collect-and-compress logic, to remove directories in cases where all children should be removed.

Then, I wondered if it would not be simpler and faster if we just remove all files and check afterwards if there are empty directories to remove. After having it implemented (in this PR), I would say it is similarly complex, but according to my tests it is faster - it only requires 0.6 to 0.9 of the time. That is because we avoid the `os.walk`.

### CODE REVIEW

File: src/poetry/installation/uninstaller.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+        # the same args, which hurts performance.
+        self._normalize_path_cached = functools.lru_cache(maxsize=None)(_normalize_path)
+        # Cache __pycache__ listings so a directory is scanned only once even
+        # though add() is called for every .py file it contains. This is safe
+        # because we are robust against files that may have been deleted
+        # in the meantime.
+        self._listdir_cached = functools.lru_cache(maxsize=None)(_safe_listdir)
+        # Normalized path of the .dist-info directory, so remove() can delete it
+        # last (see remove() for why).
+        dist_info_path: Path = dist._path  # type: ignore[attr-defined]
```
[the remark this answers is not in the document]
Comment (radoering) [reply src/poetry/installation/uninstaller.py:101]: We have already relied on `dist._path` in other parts of the code base.

File: tests/installation/test_executor.py [hunk 2]
```diff
@@ -1960,3 +1960,245 @@ def test_executor_no_supported_hash_types(
...
+def test_update_uninstalls_then_installs_when_uninstall_succeeds(
+    config: Config,
+    pool: RepositoryPool,
+    io: BufferedIO,
+    env: MockEnv,
+    mocker: MockerFixture,
+    tmp_path: Path,
+) -> None:
+    config.merge({"installer": {"builtin-uninstall": True}})
+    uninstall_mock = mocker.patch(
```
[the remark this answers is not in the document]
Comment (gmashrsa-design) [reply tests/installation/test_executor.py:2179]: > **suggestion (testing):** Consider adding a test for the update path where builtin uninstall finds nothing to remove (returns None) but the wheel install still proceeds.
> 
> Currently we only test update behavior when builtin uninstall succeeds or raises. Please also cover the case where `uninstall_distribution` returns `None` (package not installed / refused) under `builtin-uninstall=True`, and verify that `_install` still calls `WheelInstaller.install` (because `_remove` returns 0). That would complete coverage of all three builtin-uninstall outcomes for updates.

**Review by gmashrsa-design — APPROVED** [review]

File: src/poetry/installation/uninstaller.py [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+        # Finally remove the whole .dist-info directory in one call. It belongs
+        # solely to this distribution, so nothing else lives there; this also
+        # clears files an installer added without listing them in RECORD
+        # (INSTALLER, REQUESTED, direct_url.json, licenses/, ...). Done last so a
+        # failure above leaves RECORD intact for a re-run. The _permitted guard
+        # keeps the "never remove anything outside the env prefix" invariant.
+        # force=True so removal tolerates entries that vanish mid-walk (already
+        # gone is fine) and read-only files, while still raising on real errors.
+        if self._permitted(self._dist_info):
+            remove_directory(Path(self._dist_info), force=True)
```
[the remark this answers is not in the document]
Comment (radoering) [reply src/poetry/installation/uninstaller.py:215]: `force=True` triggers `_on_rm_error`, which ignores non-existing files (including the directory itself).

File: src/poetry/installation/uninstaller.py [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+        self._normalize_path_cached = functools.lru_cache(maxsize=None)(_normalize_path)
+        # Cache __pycache__ listings so a directory is scanned only once even
+        # though add() is called for every .py file it contains. This is safe
+        # because we are robust against files that may have been deleted
+        # in the meantime.
+        self._listdir_cached = functools.lru_cache(maxsize=None)(_safe_listdir)
+        # Normalized path of the .dist-info directory, so remove() can delete it
+        # last (see remove() for why).
+        dist_info_path: Path = dist._path  # type: ignore[attr-defined]
+        self._dist_info = self._normalize_path_cached(str(dist_info_path))
```
[the remark this answers is not in the document]
Comment (radoering) [reply src/poetry/installation/uninstaller.py:102]: But `remove_directory()` is only called if `self._permitted(self._dist_info)` returns `True`. Isn't this sufficient?

[the remark this answers is not in the document]
Comment (radoering) [reply src/poetry/installation/uninstaller.py:102]: A symlink pointing to another distribution's metadata _in the same_ site-packages seems to be a bit artificial. This does not really make sense because different distributions should have different metadata. I think we need only consider symlinks pointing outside of site-packages.
